### DISCUSSION TITLE
Allow an object to be treated as an expression

### DISCUSSION THREAD

**Discussion by The-Futurist** [dotnet/csharplang#6939] [Q&A]
Is there a reason why we can't write:

```cs
if (something)
{
   // do stuff with something
}
```

Where the language (when `something` is a reference type) implicitly interprets this expression as

```cs
if (something != null)
{
   // do stuff with something
}
```

**Comment by CyrusNajmabadi** [comment]
> Is there a reason why we can't write:

Yes.  We think it's inherently easy to make mistakes with that sort of truthy-ness testing.  So we made a conscious decision that the user or api author always needs to be explicit when dealing with non-booleans.

**Reply by The-Futurist** [reply]
> > Is there a reason why we can't write:
> 
> Yes. We think it's inherently easy to make mistakes with that sort of truthy-ness testing. So we made a conscious decision that the user or api author always needs to be explicit when dealing with non-booleans.

OK thanks, I can certainly understand that, but it is a shame, the code looks so much less cluttered especially when there are lots of these kinds of `if` statements.

**Reply by CyrusNajmabadi** [reply]
Fundamentally it's a design pivot for any language.  Allow brevity, but at the cost of higher incidences of mistakes?  Or have the user be more explicit, but at the cost of verbosity.  We chose the latter here based on lots of experience with languages that allow the former.  Nothing so far in the last 25 years has changed the needle there (indeed, there have been waay too many experiences with this still being an issue in other languages that has helped cement this as decision to stick with).

**Comment by The-Futurist** [comment]
This is easy enough to write though:

```cs
        public static bool exists(object? o)
        {
            return o != null;
        }
```

**Reply by CyrusNajmabadi** [reply]
Better to do `Exists<T>([NotNullWhen(true)] T? o) where T : class`.  That way you can only be called on a reference type in the first place.  No point calling this on a value type, just to box it, and then always return true.

The annotation will also ensure that flow checking knows the value is not null on teh codepaths where Exists returns 'true'.

**Reply by The-Futurist** [reply]
> Exists<T>([NotNullWhen(true)] T? o) where T : class

Thanks, that's a good way to do that. This came up as I was looking at something else, I'd be interested in your input here:

https://github.com/dotnet/csharpstandard/issues/654

**Reply by CyrusNajmabadi** [reply]
Sorry, i've got no input on htat.  You can ping rikki to see if someone can look further into that for you.

**Reply by The-Futurist** [reply]
> Sorry, i've got no input on htat. You can ping rikki to see if someone can look further into that for you.

I was just curious, from a language (not standards) perspective it seems entirely reasonable that one should be able to write that, wondered what your thoughts were.

**Reply by CyrusNajmabadi** [reply]
> wondered what your thoughts were.

Gotcha.  Thanks for the clarification.

My general intuition is that `a?.b` should be allowed if `a == null ? null : a.b` was allowed.  In the case of method groups, there is no implicit conversion with it and `null` currently.  So i'd say it should not be allowed.

Now, if someone wanted to argue that there should be an implicit conversion between `null` and a method group, i would be amenable to considering that and relaxing if we should now allow it.  If we did then allow it, i would then expect the `?.` form to work.

**Comment by HaloFour** [comment]
You can define `true`/`false` operators on your own types if you want to enable this:

```cs
public class MyClass {
    public static bool operator true(MyClass? obj) => obj != null;
    public static bool operator false(MyClass? obj) => obj == null;
}
```

Then you can use it like this:

```cs
MyClass obj = ...;
if (obj) {
    // ...
}
```

Theoretically, with [roles and extension everything](https://github.com/dotnet/csharplang/issues/5497), you could theoretically add such an operator to `System.Object`.

I'm with @CyrusNajmabadi , though, that "truthiness" as implemented this way is a vector for unexpected bugs in those languages that allows it.

**Reply by The-Futurist** [reply]
> You can define `true`/`false` operators on your own types if you want to enable this:
> 
> ```cs
> public class MyClass {
>     public static bool operator true(MyClass? obj) => obj != null;
>     public static bool operator false(MyClass? obj) => obj == null;
> }
> ```
> 
> Then you can use it like this:
> 
> ```cs
> MyClass obj = ...;
> if (obj) {
>     // ...
> }
> ```
> 
> Theoretically, with [roles and extension everything](https://github.com/dotnet/csharplang/issues/5497), you could theoretically add such an operator to `System.Object`.

That's rather interesting, thank you!
